src/features/Chat/Thread/utils.ts

```diff
@@ -249,6 +257,70 @@ export function formatChatResponse(
   }, messages);
 }
 
+function handleSubchatResponse(
+  messages: ChatMessages,
+  response: SubchatResponse,
+): ChatMessages {
+  for (let i = 0; i < messages.length; i++) {
+    const message = messages[i];
+    if (!isAssistantMessage(message)) {
+      continue;
+    }
+    if (!message.tool_calls) {
+      continue;
+    }
+    for (let j = 0; j < message.tool_calls.length; j++) {
+      const tool_call = message.tool_calls[j];
+      if (tool_call.id !== response.tool_call_id) {
+        continue;
+      }
+
+      const newMessage = { ...message };
+      const tool_calls = [...message.tool_calls];
+      newMessage.tool_calls = tool_calls;
+
+      tool_calls[j] = { ...tool_call };
+      tool_calls[j].subchat = response.subchat_id;
+
+      const { add_message } = response;
+      if (isSubchatContextFileResponse(add_message)) {
+        const content = parseOrElse<ChatContextFile[]>(add_message.content, []);
+        const attached_files = tool_calls[j].attached_files ?? [];
+        tool_calls[j].attached_files = [...attached_files];
+        for (const file of content) {
+          tool_calls[j].attached_files?.push(file.file_name);
+        }
+      }
+
+      const res = [...messages];
+      res[i] = newMessage;
+      return res;
+    }
+  }
+  return messages;
```

Comment:
Usually when I have to do some iterative task,  I'll try to either use array methods like `filter`,`find`, `map` ect,
or use a recursive approach.

```js
function handleSubchatResponse(
  messages: ChatMessages,
  response: SubchatResponse,
): ChatMessages {
  function iter(msgs: ChatMessages, response: SubchatResponse, memo: ChatMessages = []) {
    // continue / exit conditions
     if(msgs.length === 0) return memo;
     const [head, ...tail] = msgs;
     if(!isAssistantMessage(head) || !head.tool_calls) return iter(tail, response, memo)
     const maybeToolCall = head.tool_call.find(toolCall => toolCall.id === response.tool_call_id)
     if(!maybeToolCall) return iter(tail, response, memo);
     
     // logic
     const toolCallWithCubChat = {
        ...maybeToolCall,
        subchat: response.subchat_id,
     }
    ....
    // next iteration
     const nextMemo = [...memo, toolCallWithSubChat]
     return iter(tail, response, nextMemo)
  }
  
  return iter(messages, response)
}
```